provisioning/terraform/cli_dist_cloudfront.tf

```diff
@@ -0,0 +1,78 @@
+// https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/cloudfront_distribution
+
+locals {
+  cli_dist_default_origin_id = "cli_dist_default_origin"
+}
+
+// 88a5eaf4-2fd4-4709-b370-b4c650ea3fcf
+data "aws_cloudfront_origin_request_policy" "aws_managed_origin_request_policy_cors_s3_origin" {
+  name = "Managed-CORS-S3Origin"
+}
+
+resource "aws_cloudfront_distribution" "cli_dist_cloudfront" {
+  depends_on = [
+    //@TODO aws_acm_certificate.keboola_assets_acm_certificate,
+    aws_s3_bucket.cli_dist_bucket
+  ]
+  aliases = [
+    //@TODO var.hosted_zone_subdomain_name
+  ]
+  enabled         = true
+  http_version    = "http2"
+  is_ipv6_enabled = true
+  default_cache_behavior {
+    cache_policy_id          = aws_cloudfront_cache_policy.cli_dist_cloudfront_cache_policy.id
+    compress                 = false
+    origin_request_policy_id = data.aws_cloudfront_origin_request_policy.aws_managed_origin_request_policy_cors_s3_origin.id
+    cached_methods           = ["GET", "HEAD"]
+    target_origin_id         = local.cli_dist_default_origin_id
+    viewer_protocol_policy   = "https-only"
+    allowed_methods          = ["GET", "HEAD"]
+  }
+  origin {
+    domain_name = aws_s3_bucket.cli_dist_bucket.bucket_regional_domain_name
+    origin_id   = local.cli_dist_default_origin_id
+    s3_origin_config {
+      origin_access_identity = aws_cloudfront_origin_access_identity.cli_dist_cloudfront_origin_access_identity.cloudfront_access_identity_path
+    }
+  }
+  viewer_certificate {
+    //@TODO acm_certificate_arn      = aws_acm_certificate.keboola_assets_acm_certificate.arn
+    cloudfront_default_certificate = true
```

Comment:
@eddycek @Halama Můžu teda tady nechat defaultní certifikát s tím, že ten správnej se společně s aliasem nakliká už manuálně, jo?

Jinak mám hotový TF bakend a tady ty potřebný resoursy vesměs taky. Mám to deploynutý do testovacího AWS accountu ručně. Teď to budu přidávat do CI.

Reply:
Viz. https://d4yf8q7smaed8.cloudfront.net/index.html -> http://ci-cli-dist-keboola-com.s3-website-us-east-1.amazonaws.com/ (Ten index vyhazuje chybu, protože je tam třeba zahardcodovat url bucketu: https://github.com/keboola/keboola-as-code/blob/0a2ea68ea147077de439f8f2379062db30bcfb56/provisioning/terraform/index.html#L13-L14) To asi v Terraformu moc nenaskriptuju. Nebo jo? 🙂 (Je to javascript co emuluje file server, jako je tady: https://cli-dist.keboola.com/)

Reply:
Co se týče certifikátu vytvoříme a zvalidujeme ho ručně. V tom Assets CDN byla přímo v tom accountu i managovaná doména takže to šlo automaticky https://github.com/keboola/keboola-assets-cdn/blob/main/terraform/keboola_assets_acm_certificate.tf to tady nepůjde.

ARN toho certifikátu zadáme jako variable, takže se bude referencovat:
```
 acm_certificate_arn      = var.aws_acm_certificatearn
```

`cloudfront_default_certificate` se nastavovat nebude

Reply:
ok 👍

Reply:
| To asi v Terraformu moc nenaskriptuju. Nebo jo? 🙂
Na to mrknu. Spíš si říkám jestli nebude problém že bude potom ukazovat na jiný bucket? Myslel jsem že to je celé schované za CDN, ale tady https://cli-dist.keboola.com/?prefix=apk/ např vidím že to odkazuje přímo na S3 https://keboola-as-code-dist.s3.amazonaws.com/apk/keboola.rsa.pub to je záměrně?

Reply:
ale přes CDN zároveň funguje https://cli-dist.keboola.com/apk/keboola.rsa.pub

Reply:
No ten bucket bych právě potřeboval parametrizovat. 😄 A jo, ty url linky tam generuje ta javascriptová libka a ta pracuje přímo s S3 bucketem. Podívám se jestli jí nejde podstrčit jiný base url.

Reply:
Myslim ze by to mohlo jit replacnout pomoci https://www.terraform.io/language/functions/templatefile zavolat tady https://github.com/keboola/keboola-as-code/pull/872/files#diff-a8ac8ea767d3ef2b6f3fa3f35ba06c0ed5bfdc0e2db32b165ac31a3d85816647R57 Ale asi jsme to jeste nikde nepouzivali

Reply:
Me slo ale jinak o to proc je vubec public ten bucket, proc to cele neni za tou CDN.

Reply:
Jasně, myslím že to není třeba. Resp. tak maximálně kvůli tý JS libce na procházení bucketu. Já jsem to zatím zkopíroval z toho legacy S3+CF abych to vůbec rozjel. Můžu to upravit.

Reply:
jo jasne zacneme tim jak to funguje ted a az to pojede na testingu alespon muze se jeste doladit.

Reply:
Pro ten testing bych to nahodil pod `https://cli-dist.keboola.dev/`. Připravím tam ten ACM, teď to na něm padlo.

Reply:
Založil jsem a zvalidoval ten cert https://us-east-1.console.aws.amazon.com/acm/home?region=us-east-1#/certificates/30e7ec1d-c7e9-4d9e-a8be-4d03d9f22ee7

Jeho ARN pro testing do vars je tedy: `arn:aws:acm:us-east-1:813746015128:certificate/30e7ec1d-c7e9-4d9e-a8be-4d03d9f22ee7`

Reply:
koukám že to asi prošlo https://github.com/keboola/keboola-as-code/actions/runs/3226554074/jobs/5280214348
Nasměruju teď ještě ručně DNS https://cli-dist.keboola.dev/ na tu nově vytvořenou distribuci https://us-east-1.console.aws.amazon.com/cloudfront/v3/home?region=us-east-1#/distributions/E3PZZ51R71GEQP

Reply:
@JakubMatejka Tak ještě aby fungoval custom domain name bude potřeba nastavit pro tu CF distrubuci alias stejně jako tady https://github.com/keboola/keboola-assets-cdn/blob/0bddd1b5835c7c91dcdf4ef7fa19a42b664eb7e9/terraform/ui_keboola_assets_distribution.tf#L29 a https://github.com/keboola/keboola-assets-cdn/blob/0269c45091e7b206bc6d1ebac0f5566eee5e950d/terraform/testing.tfvars#L3

A do variable dat pro testing `cli-dist.keboola.dev`

Reply:
Hotovo, funfuje: https://cli-dist.keboola.dev/

Reply:
Super, jenom to automaticky nezobrazí ten index ale to asi ješě půjde poladit https://cli-dist.keboola.dev/index.html

Reply:
Jo jo, to teď doladím (společně s funkcionalitou toho index.html, zatím to ještě nejede jak by mělo).

Reply:
Pak až to pojede tím současným způsobem můžeme mrknout jestli tu S3 nepůjde úplně schovat aby to bylo dostupné jen přes CDN.

Reply:
Ha, v téhle kombinaci už ten skript funguje a nabízí k downloadu i linky přes CDN: http://ci-cli-dist-keboola-com.s3-website-us-east-1.amazonaws.com/?prefix=msi/ (na https://cli-dist.keboola.dev/index.html to ale ještě není vidět, protože ta CDNka to nejspíš cachuje). Zkusím to zparametrizovat přes tu `templatefile` 

![CleanShot 2022-10-11 at 15 03 49](https://user-images.githubusercontent.com/215660/195098492-bf0e1da5-ce99-4837-ab7a-6372d3047498.jpg)

Reply:
Tak vyřešeno včetně toho zobrazení indexu na https://cli-dist.keboola.dev/ (https://github.com/keboola/keboola-as-code/pull/872/commits/8ad3badb9853573a83e948481de5f227498b063b)

Reply:
Deploy TF na testing funguje. Vytvořím teda TF backend v produkčním účtu, aby procházel i ten Plán pro Produkci.

![CleanShot 2022-10-11 at 15 35 58@2x](https://user-images.githubusercontent.com/215660/195105928-1528f2b8-2e29-47d5-b24c-7ac99bb51030.jpg)

Reply:
super, to vypadá dobře.

Reply:
jj 👍

Reply:
Připravil jsem https://github.com/keboola/keboola-as-code/pull/876/files

Reply:
Když jsem se na to díval ještě si nejsem úplně jistý tou strukturou a namingem v tom provisioningu 
<img width="1356" alt="image" src="https://user-images.githubusercontent.com/903531/195123770-e2c7b4cd-da11-421a-83fa-3d12c010eaf7.png">

Myslím že by to mělo být zanořeno podle "use casu" protože ty resourcy co tam jsou ted jako aws a azure se časem možná také switchnou do terraform a jde o templates-api které se nasazuje do kbc stacků narozdíl od cli distribuce kterou řešíme teď.

Myslím že by to pak mohlo vypadat:
- `provisioning/templates-api` 
- `provisioning/cli-dist` (neboo případně kac-assets)
- `provisioinig/etcd`

Možná by se teda už teď mohl ten nově přidaný terraform folder přejmenovat na ten `provisioning/cli-dist` nebo jinou variantu.
Snad je to srozumitelný, kdyžtak vysvětlím více.

Reply:
Jo jo, rozumím. Moc jsem právě nevěděl co s tím. Ok 👍